lib/APPTurtle/src/CustomRosTransport.cpp

```diff
@@ -0,0 +1,211 @@
+ * Local Variables
+ *****************************************************************************/
+
+/** Instance of the UDP client */
+static WiFiUDP udpClient;
+
+/******************************************************************************
+ * Public Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Protected Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Private Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * External Functions
+ *****************************************************************************/
+
+bool custom_transport_open(uxrCustomTransport* transport)
+{
+    bool isOpen = false;
+
+    if (nullptr == transport)
+    {
+        LOG_ERROR("Transport is nullptr.");
+    }
+    else
+    {
+        const micro_ros_agent_locator* locator = reinterpret_cast<micro_ros_agent_locator*>(transport->args);
+        const int                      UDP_OK  = 1;
+
+        if (UDP_OK != udpClient.begin(locator->port))
+        {
+            LOG_ERROR("UDP begin error");
+        }
+        else
+        {
+            isOpen = true;
+        }
+    }
+
+    return isOpen;
+}
+
+bool custom_transport_close(uxrCustomTransport* transport)
+{
+    UTIL_NOT_USED(transport);
+
+    udpClient.stop();
+    return true;
+}
+
+size_t custom_transport_write(uxrCustomTransport* transport, const uint8_t* buffer, size_t size, uint8_t* errorCode)
+{
+    size_t sent = 0;
+
+    if ((nullptr == transport) || (nullptr == buffer) || (0 == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        const micro_ros_agent_locator* locator = reinterpret_cast<micro_ros_agent_locator*>(transport->args);
+        const int                      UDP_OK  = 1;
+        int                            ret     = UDP_OK;
+
+        ret = udpClient.beginPacket(locator->address, locator->port);
+
+        if (UDP_OK != ret)
+        {
+            LOG_ERROR("UDP beginPacket error");
+        }
+        else
+        {
+            size_t bytesToWrite = udpClient.write(buffer, size);
+
+            if (bytesToWrite != size)
+            {
+                LOG_ERROR("UDP write error");
+            }
+            else
+            {
+                ret = udpClient.endPacket();
+
+                if (UDP_OK != ret)
+                {
+                    LOG_ERROR("UDP endPacket error");
+                }
+                else
+                {
+                    sent = bytesToWrite;
+                }
+            }
+        }
+
+        if (UDP_OK != ret)
```

Comment:
if you fail already at line 131, you will enter this if here again with providing an error. Is this intended ?

Reply:
Yes it is intended. Same applies if it fails in line 147. In these two cases it is possible to get the write error from the UDP client